grpc/execution/server.go

```diff
@@ -123,11 +129,232 @@ func (s *ExecutionServiceServer) FinalizeBlock(ctx context.Context, req *executi
+	executionv1a2.UnimplementedExecutionServiceServer
+
+	eth *eth.Ethereum
+	bc  *core.BlockChain
+}
+
+func NewExecutionServiceServerV1Alpha2(eth *eth.Ethereum) *ExecutionServiceServerV1Alpha2 {
+	bc := eth.BlockChain()
+
+	return &ExecutionServiceServerV1Alpha2{
+		eth: eth,
+		bc:  bc,
+	}
+}
+
+// GetBlock will return a block given an identifier.
+func (s *ExecutionServiceServerV1Alpha2) GetBlock(ctx context.Context, req *executionv1a2.GetBlockRequest) (*executionv1a2.Block, error) {
+	log.Info("GetBlock called request", "request", req)
+
+	res, err := s.getBlockFromIdentifier(req.GetIdentifier())
+	if err != nil {
+		return nil, status.Error(codes.NotFound, "Block header cannot be converted to execution block")
+	}
+
+	return res, nil
+}
+
+// BatchGetBlocks will return an array of Blocks given an array of block
+// identifiers.
+func (s *ExecutionServiceServerV1Alpha2) BatchGetBlocks(ctx context.Context, req *executionv1a2.BatchGetBlocksRequest) (*executionv1a2.BatchGetBlocksResponse, error) {
+	var blocks []*executionv1a2.Block
+
+	ids := req.GetIdentifiers()
+	for _, id := range ids {
+		block, err := s.getBlockFromIdentifier(id)
+		if err != nil {
+			return nil, err
+		}
+
+		blocks = append(blocks, block)
+	}
+
+	res := &executionv1a2.BatchGetBlocksResponse{
+		Blocks: blocks,
+	}
+
+	return res, nil
+}
+
+// ExecuteBlock drives deterministic derivation of a rollup block from sequencer
+// block data
+func (s *ExecutionServiceServerV1Alpha2) ExecuteBlock(ctx context.Context, req *executionv1a2.ExecuteBlockRequest) (*executionv1a2.Block, error) {
+	log.Info("ExecuteBlock called request", "request", req)
+
+	// Validate block being created has valid previous hash
+	prevHeadHash := common.BytesToHash(req.PrevBlockHash)
+	softHash := s.bc.CurrentSafeBlock().Hash()
+	if prevHeadHash != softHash {
+		return nil, status.Error(codes.FailedPrecondition, "Block can only be created on top of soft block.")
+	}
+
+	// The Engine API has been modified to use transactions from this mempool and
+	// abide by it's ordering.
+	s.eth.TxPool().SetAstriaOrdered(req.Transactions)
+
+	// Build a payload to add to the chain
+	payloadAttributes := &miner.BuildPayloadArgs{
+		Parent:       prevHeadHash,
+		Timestamp:    uint64(req.GetTimestamp().GetSeconds()),
+		Random:       common.Hash{},
+		FeeRecipient: common.Address{},
+	}
+	payload, err := s.eth.Miner().BuildPayload(payloadAttributes)
+	if err != nil {
+		log.Error("failed to build payload", "err", err)
+		return nil, status.Error(codes.InvalidArgument, "could not build block with provided txs")
+	}
+
+	// call blockchain.InsertChain to actually execute and write the blocks to
+	// state
+	block, err := engine.ExecutableDataToBlock(*payload.Resolve().ExecutionPayload)
+	if err != nil {
+		return nil, err
+	}
+	blocks := types.Blocks{
+		block,
+	}
+	n, err := s.bc.InsertChain(blocks)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "failed to insert block to chain")
```

Comment:
wrap error

Reply:
Don't want to expose internal errors via API, this is purposefully non-wrapped.

Reply:
okay, that'll make it hard to debug though if there are any errors, as the error is now completely ignored. maybe log it?